includes/admin/class-wc-rest-payments-reader-controller.php

```diff
@@ -62,4 +100,69 @@ public function get_summary( $request ) {
 		return rest_ensure_response( $summary );
 	}
 
+	/**
+	 * Renders HTML for a print receipt
+	 *
+	 * @param WP_REST_Request $request Full data about the request.
+	 * @return WP_HTTP_RESPONSE|WP_Error
+	 */
+	public function generate_print_receipt( $request ) {
```

Comment:
This is very safe, but difficult to read. Let's address other comments and I'll suggest a simplified workflow here on the next review iteration.

Reply:
Here we go:

```php
		try {
			/* Collect the data, available on the server side. */
			$payment_intent = $this->api_client->get_intent( $request->get_param( 'payment_id' ) );
			if ( 'succeeded' !== $payment_intent->get_status() ) {
				throw new \RuntimeException( __( 'Invalid payment', 'woocommerce-payments' ) );
			}
			$charge = $this->api_client->get_charge( $payment_intent->get_charge_id() );
			
			/* Collect receipt data, stored on the store side. */
			$order = wc_get_order( $charge['order']['number'] );
			if ( $order === false ) {
				throw new \RuntimeException( __( 'Order not found', 'woocommerce-payments' ) );
			}
			$settings = [
				'business_name' => $this->wcpay_gateway->get_option( 'account_business_name' ),
				'support_info'  => [
					'address' => $this->wcpay_gateway->get_option( 'account_business_support_address' ),
					'phone'   => $this->wcpay_gateway->get_option( 'account_business_support_phone' ),
					'email'   => $this->wcpay_gateway->get_option( 'account_business_support_email' ),
				],
			];
			$receipt_data = $this->receipts_service->get_receipt_markup( $settings, $order, $charge );
		} catch ( \Throwable $e ) {
			$error_status_code = $e instanceof API_Exception ? $e->get_http_code() : 500;
			return rest_ensure_response( new WP_Error( 'generate_print_receipt_error', $e->getMessage(), [ 'status' => $error_status_code ] ) );
		}

		/**
		 * WP_REST_Server will convert the response data to JSON prior to output it.
		 * Using this filter to prevent it, and output the data from WP_HTTP_Response instead.
		 */
		add_filter(
			'rest_pre_serve_request',
			function ( bool $served, WP_HTTP_Response $response ) : bool {
				echo $response->get_data(); // @codingStandardsIgnoreLine
				return true;
			},
			10,
			2
		);

		return new WP_HTTP_Response(
			$receipt_data,
			200,
			[ 'Content-Type' => 'text/html; charset=UTF-8' ]
		);
```

Reply:
Thanks @kalessil better this way

Reply:
@kalessil Why are you catching `Throwable` over `Exception`? Do we really need to worry about `Error` being thrown?

Reply:
@KarlisJ: to avoid uncontrolled crashes. This endpoint to be used by the Mobile app and making it as reliable as possible is a good thing.